website/blog/2024-08-07-manipulate-devtools.md

```diff
@@ -0,0 +1,325 @@
+  mochaOpts: {
+    ui: "bdd",
+    timeout: 1000000,
+  },
+};
+```
+
+In this configuration, we specify the Chrome browser with options to auto-open DevTools for tabs and load the PixiJS DevTools extension. The `extensions` option reads the extension file in base64 format.
+We need also the `--auto-open-devtools-for-tabs` flag to automatically open the DevTools (Developer Tools) panel.
+
+## Writing the Test Script
+
+Next, we write a test script to interact with the PixiJS extension through Chrome DevTools. Create a test file, for example, `devtools.test.js`, and add the following code:
+
+```javascript
+import { $, browser } from "@wdio/globals";
+
+describe("DevTools Test on latest chrome", function () {
+  it("Devtools Navigation", async function () {
+    // Open an example game.
+    await browser.url(
+      "https://pixijs.io/examples-v5/#/demos-basic/container.js"
+    );
+
+    // Get window handles.
+    const handles = await browser.getWindowHandles();
+
+    // Switch to devtools window.
+    await browser.switchToWindow(handles[1]);
+
+    // Navigate through devtool tabs (for Linux use ['Control', '[']).
+    await browser.keys(["Meta", "["]);
+
+    // Switch to devtool extension iframe.
+    await browser.switchToFrame(0);
+
+    // Manipulate extension.
+    await $(
+      "body > div > div > div.status.svelte-iqeeoq > div.patch.svelte-iqeeoq > button"
+    ).click();
+    await $(
+      "body > div > div > div.outliner.svelte-iqeeoq > div > div.body.svelte-1vjyr8f > div:nth-child(1) > button:nth-child(4)"
+    ).click();
+    await $(
+      "body > div > div > div.outliner.svelte-iqeeoq > div > div.body.svelte-1vjyr8f > div:nth-child(1) > button:nth-child(4)"
+    ).click();
+
+    // Switch back to parent frame.
+    await browser.switchToParentFrame();
+
+    // Switch back to main window.
+    await browser.switchToWindow(handles[0]);
+  });
+});
+```
+
+### Explanation of the Script
+
+1. **Open a URL**: The script navigates to an example PixiJS game.
+2. **Get Window Handles**: Retrieves the window handles for the main window and the DevTools window.
+3. **Switch to DevTools Window**: Switches context to the DevTools window using its handle.
+4. **Navigate DevTools Tabs**: Uses keyboard shortcuts to navigate through DevTools tabs. Note: For Linux, replace `['Meta', '[']` with `['Control', '[']`.
+5. **Switch to Iframe**: Switches to the iframe that contains the PixiJS extension.
+6. **Interact with the Extension**: Performs clicks on specific elements within the extension.
+7. **Switch Back**: Returns to the parent frame and then back to the main window.
+
+## Running the Test
+
+To run the test, execute the following command:
+
+```sh
+npx wdio run wdio.conf.js
+```
+
+This will launch Chrome, open the specified URL, interact with the DevTools, and manipulate the PixiJS extension as defined in the test script.
+
+#### Demonstration
+![DevTools v8](assets/DevTools_v8.gif)
+
+# Devtools in WebdriverIO v7
```

Comment:
It seems like a lot of content here is very similar. How about we only keep on section and for the code use Docusaurus [tabs](https://docusaurus.io/docs/markdown-features/tabs) to show the example using WebDriver primitives vs. Puppeteer to work with devtools

Reply:
We could use that but the issue would be that the wdio v8 uses a version of Puppeteer that there is no option to manipulate devtools as it doesn't give you the choice to change the type of devtools from `other` to `page`. We achieve that by forcing `puppeteer-core` to the latest version, where there is a new option to set a target `asPage()`  https://pptr.dev/api/puppeteer.target.aspage/ . Do you prefer to just write about v8 using WebDriver primitives?

Reply:
done